src/null-safety/understanding-null-safety/index.md

```diff
@@ -0,0 +1,1410 @@
+}
+```
+
+Spot the bug? The `.where()` method is lazy, so it returns an `Iterable`, not a
+`List`. This program compiles but then throws an exception at runtime when it
+tries to cast that `Iterable` to the `List` type that `filterEvens` declares it
+returns. With the removal of implicit downcasts, this becomes a compile error.
+
+Where were we? Right, OK, so it's as if we've taken the universe of types in
+your program and split them into two halves:
+
+<img src="understanding-null-safety/bifurcate.png">
+
+There is a region of non-nullable types. Those types let you access all of the
+interesting methods, but can never ever contain `null`. And then there is a
+parallel family of all of the corresponding nullable types. Those permit `null`,
+but you can't do much with them. We let values flow from the non-nullable side
+to the nullable side because doing so is safe, but not the other direction.
+
+That seems like nullable types are basically useless. They have no methods and
+you can't get away from them. Don't worry, we have a whole suite of features to
+help you move values from the nullable half over to the other side that we will
+get to soon.
+
+### Top and bottom
+
+This section is a little esoteric. You can mostly skip it, except for two
+bullets at the very end, unless you're into type system stuff. Imagine all the
+types in your program with edges between ones that are subtypes and supertypes
+of each other. If you were to draw it, like the diagrams in this doc, it would
+form a huge directed graph with supertypes like `Object` near the top and leaf
+classes like your own types near the bottom.
+
+If that directed graph comes to a point at the top where there is a single type
+that is the supertype (directly or indirectly), that type is called the *top
+type*. Likewise, if there is a weird type at that bottom that is a subtype of
+every type, you have a *bottom type*. (In this case, your directed graph is a
+[lattice][].)
+
+[lattice]: https://en.wikipedia.org/wiki/Lattice_(order)
+
+It's convenient if your type system has a top and bottom type, because it means
+that type-level operations like least upper bound (which type inference uses to
+figure out the type of a conditional expression based on the types of its two
+branches) can always produce a type. Before null safety, `Object` was Dart's top
+type and `Null` was its bottom type.
+
+Since `Object` is non-nullable now, it is no longer a top type. `Null` is not a
+subtype of it. Dart has no *named* top type. If you need a top type, you want
+`Object?`. Likewise, `Null` is no longer the bottom type. If it was, everything
+would still be nullable. Instead, we've added a new bottom type named `Never`:
+
+<img src="understanding-null-safety/top-and-bottom.png" width="50%">
+
+In practice, this means:
+
+*   If you want to indicate that you allow a value of any type, use `Object?`
+    instead of `Object`. In fact, it becomes pretty rare to use `Object` since
```

Comment:
Do we have data on the relative usage of `Object?` and `Object` ? I'm suspicious that `Object` is less rare than indicated here.

Reply:
No official data. This is based on my experience migrating the core libraries and tests. Most places that were previously using `Object` ended up getting changed to `Object?`. The only exceptions I can think of offhand are code dealing with errors/exceptions because the language specifically prohibits throwing `null`, and overrides of `==` because, again, the language prohibits `null`.

Reply:
I wouldn't consider the core libraries and tests to be particularly representative of the ecosystem.

I'm worried about sending an unintentionally prescriptive message with this.